DIPs/13NNN-WGB.md

```diff
@@ -0,0 +1,893 @@
+}
+```
+
+**AndAnd/OrOr expressions**. *e1 || e2* and *e1 && e2* expressions are treated the following way:
+
+  * if `e1` accesses `x` and `e2` does not, then `e1` is the last access of `x`;
+  * if `e2` accesses `x` and `e1` does not, then `e2` is the last access of `x`;
+  * if both `e1` and `e2` access `x`, then `e2` is the last access of `x`;
+
+**Partial move**. A variable of an aggregated type (struct or class) may contain an EMO field:
+
+```
+void fun(T t);
+void gun(S s);
+struct S
+{
+    T t;    // T is EMO
+}
+{
+    S s;
+    fun(s.t);
+    gun(s);
+}
+```
+
+Although the call to `fun` is the last time `s.t` is accessed, `s` is still used after that. To correctly
+treat this situation, it is necessary, that `s.t` will make a copy, otherwise when `gun` is called `s` will
+be in a partially invalid state (the move could potentially alter `s.t`). However, if `fun` would be the last
+access of `s`, then it would be no problem to move `t`. Therefore, the generalized rule is that an access to
+an EMO field of an aggregate will be moved only if that is the last access of the containing variable.
+
+**Pointers**. Whenever the address of a variable `x` is taken, `x` will lose the possibility of last access optimization.
+For example:
+
+```
+{
+  S s;
+  int* p = &s.i;
+  func(s);  // not last use of s
+  *p = 3;   // because s is still being written to
+}
+{
+  ...
+  S s;
+  S* ps = &s;
+  func(s);      // not last use
+  int j = ps.i; // because s is still being read
+}
+{
+  ...
+  S s;
+  S* ps = &s;
+  func(s);   // ps still points to s, so it's not last use, even if ps is never used
+}
+
+struct T { int j; S s; ~this(); }
+
+{
+  T t;
+  func(t.s);  // not last use because of T's destructor
+}
+```
+
+An expression yielding an rvalue of an EMO is always the Last Use of that rvalue,
+and a move is done.
+
+```
+void func(S);
+...
+func(S());  // S() is an rvalue, so always a move
+```
+
+
+### Destruction
+
+Moving an EMO means that the responsiblity for destruction of that value
+moves along with it.
+
+```
+S test(S s)
+{
+    return func(s); // destruction of s is now func's responsibility
+}
+```
+
+
+For merging control paths where one path may have moved the lvalue:
+
+```
+{
+    S s;
+    if (cond)
+        func(s); // copy or move?
+    s.__dtor();  // but what if s was moved?
+}
+```
+the implementation may choose to use a copy, or can use a move with a flag
+that gates the destructor call:
+```
+{
+    S s;
+    bool flag = true;
+    if (cond)
+    {
+        func(s); // copy or move?
+        flag = false;
+    }
+    flag && s.__dtor();
+}
+```
+If `func()` may throw an exception, `flag` must be set to false before
+`func()` is called.
+
+#### Assignment after Move
+
+```
+S s, t;
+func(s); // A
+s = t;   // B
+```
+may be compiled two ways:
+
+1. A: `s` is copied, B: `s` is assigned
+2. A: `s` is moved, B: `s` gets constructed, not assigned
+
+at the implementation's discretion. Case 1 can be done for a quick build,
```

Comment:
But this should compile:
```d
struct S {
    // non-copyable
    @disable this(ref S);
    // EMO
    this(S other) { /* ... */ }
    void opAssign(S other) { /* ... */ }
}

S createNewS();
void consume(S);

void usage() {
    S s = createNewS();
    consume(s); // cannot be copied
    s = createNewS();
    consume(s); // cannot be copied
    // s is not destructed here
}
```
This precludes such discretionary power of the implementation. It would **have to** treat pass before a `=` as a move-then-construct. But, if so defined, this may make `=` a bit confusing:
```d
void usage2() {
    auto s = createNewS();
    s = createNewS(); // move-assign
    consume(s); // move
    s = createNewS(); // move-construct
    // s is destructed here
}
```